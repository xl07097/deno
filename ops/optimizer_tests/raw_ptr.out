#[allow(non_camel_case_types)]
///Auto-generated by `deno_ops`, i.e: `#[op]`
///
///Use `op_ffi_ptr_of::decl()` to get an op-declaration
///you can include in a `deno_core::Extension`.
pub struct op_ffi_ptr_of;
#[doc(hidden)]
impl op_ffi_ptr_of {
    pub const fn name() -> &'static str {
        stringify!(op_ffi_ptr_of)
    }
    #[allow(clippy::not_unsafe_ptr_arg_deref)]
    pub extern "C" fn v8_fn_ptr<'scope, FP>(
        info: *const deno_core::v8::FunctionCallbackInfo,
    )
    where
        FP: FfiPermissions + 'static,
    {
        let info = unsafe { &*info };
        let scope = &mut unsafe { deno_core::v8::CallbackScope::new(info) };
        let args = deno_core::v8::FunctionCallbackArguments::from_function_callback_info(
            info,
        );
        let rv = deno_core::v8::ReturnValue::from_function_callback_info(info);
        Self::v8_func::<FP>(scope, args, rv);
    }
    pub const fn decl<'scope, FP>() -> deno_core::OpDecl
    where
        FP: FfiPermissions + 'static,
    {
        deno_core::OpDecl {
            name: Self::name(),
            v8_fn_ptr: Self::v8_fn_ptr::<FP> as _,
            enabled: true,
            fast_fn: {
                use deno_core::v8::fast_api::Type::*;
                use deno_core::v8::fast_api::CType;
                Some(
                    deno_core::v8::fast_api::FastFunction::new(
                        &[
                            V8Value,
                            TypedArray(CType::Uint8),
                            TypedArray(CType::Uint32),
                            CallbackOptions,
                        ],
                        CType::Void,
                        op_ffi_ptr_of_fast_fn::<FP> as *const ::std::ffi::c_void,
                    ),
                )
            },
            is_async: false,
            is_unstable: false,
            is_v8: false,
            force_registration: false,
            arg_count: 2usize as u8,
        }
    }
    #[inline]
    #[allow(clippy::too_many_arguments)]
    fn call<FP>(state: &mut OpState, buf: *const u8, out: &mut [u32])
    where
        FP: FfiPermissions + 'static,
    {}
    pub fn v8_func<'scope, FP>(
        scope: &mut deno_core::v8::HandleScope<'scope>,
        args: deno_core::v8::FunctionCallbackArguments,
        mut rv: deno_core::v8::ReturnValue,
    )
    where
        FP: FfiPermissions + 'static,
    {
        let ctx = unsafe {
            &*(deno_core::v8::Local::<deno_core::v8::External>::cast(args.data()).value()
                as *const deno_core::_ops::OpCtx)
        };
        let arg_0 = {
            let value = args.get(0usize as i32);
            match deno_core::v8::Local::<deno_core::v8::ArrayBuffer>::try_from(value) {
                Ok(b) => {
                    if let Some(data) = b.data() {
                        data.cast::<u8>().as_ptr()
                    } else {
                        std::ptr::null::<u8>()
                    }
                }
                Err(_) => {
                    if let Ok(view)
                        = deno_core::v8::Local::<
                            deno_core::v8::ArrayBufferView,
                        >::try_from(value) {
                        let offset = view.byte_offset();
                        let buffer = match view.buffer(scope) {
                            Some(v) => v,
                            None => {
                                return deno_core::_ops::throw_type_error(
                                    scope,
                                    format!("Expected ArrayBufferView at position {}", 0usize),
                                );
                            }
                        };
                        let store = if let Some(data) = buffer.data() {
                            data.cast::<u8>().as_ptr()
                        } else {
                            std::ptr::null_mut::<u8>()
                        };
                        unsafe { store.add(offset) }
                    } else {
                        return deno_core::_ops::throw_type_error(
                            scope,
                            format!("Expected ArrayBufferView at position {}", 0usize),
                        );
                    }
                }
            }
        };
        let arg_1 = if let Ok(view)
            = deno_core::v8::Local::<
                deno_core::v8::Uint32Array,
            >::try_from(args.get(1usize as i32)) {
            let (offset, len) = (view.byte_offset(), view.byte_length());
            let buffer = match view.buffer(scope) {
                Some(v) => v,
                None => {
                    return deno_core::_ops::throw_type_error(
                        scope,
                        format!("Expected Uint32Array at position {}", 1usize),
                    );
                }
            };
            if let Some(data) = buffer.data() {
                let store = data.cast::<u8>().as_ptr();
                unsafe {
                    ::std::slice::from_raw_parts_mut(
                        store.add(offset) as *mut u32,
                        len / 4,
                    )
                }
            } else {
                &mut []
            }
        } else {
            return deno_core::_ops::throw_type_error(
                scope,
                format!("Expected Uint32Array at position {}", 1usize),
            );
        };
        let result = Self::call::<
            FP,
        >(&mut std::cell::RefCell::borrow_mut(&ctx.state), arg_0, arg_1);
        let op_state = ::std::cell::RefCell::borrow(&*ctx.state);
        op_state.tracker.track_sync(ctx.id);
    }
}
#[allow(clippy::too_many_arguments)]
fn op_ffi_ptr_of_fast_fn<'scope, FP>(
    _: deno_core::v8::Local<deno_core::v8::Object>,
    buf: *const deno_core::v8::fast_api::FastApiTypedArray<u8>,
    out: *const deno_core::v8::fast_api::FastApiTypedArray<u32>,
    fast_api_callback_options: *mut deno_core::v8::fast_api::FastApiCallbackOptions,
) -> ()
where
    FP: FfiPermissions + 'static,
{
    use deno_core::v8;
    use deno_core::_ops;
    let __opts: &mut v8::fast_api::FastApiCallbackOptions = unsafe {
        &mut *fast_api_callback_options
    };
    let __ctx = unsafe {
        &*(v8::Local::<v8::External>::cast(unsafe { __opts.data.data }).value()
            as *const _ops::OpCtx)
    };
    let state = &mut ::std::cell::RefCell::borrow_mut(&__ctx.state);
    let buf = unsafe { (&*buf).get_storage_if_aligned().unwrap_unchecked() }.as_ptr();
    let out = match unsafe { &*out }.get_storage_if_aligned() {
        Some(v) => v,
        None => {
            unsafe { &mut *fast_api_callback_options }.fallback = true;
            return Default::default();
        }
    };
    let result = op_ffi_ptr_of::call::<FP>(state, buf, out);
    result
}
