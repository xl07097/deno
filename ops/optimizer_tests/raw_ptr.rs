fn op_ffi_ptr_of<FP>(state: &mut OpState, buf: *const u8, out: &mut [u32])
where
  FP: FfiPermissions + 'static,
{
  // ..
}
