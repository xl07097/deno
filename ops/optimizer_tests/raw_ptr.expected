=== Optimizer Dump ===
returns_result: false
has_ref_opstate: true
has_rc_opstate: false
has_fast_callback_option: false
needs_fast_callback_option: true
fast_result: Some(Void)
fast_parameters: [V8Value, Uint8Array, Uint32Array]
transforms: {1: Transform { kind: PtrU8, index: 1 }, 2: Transform { kind: SliceU32(true), index: 2 }}
is_async: false
fast_compatible: true
